docs: add "Immutability Helpers" recipe (#1252)
ngxs/store: docs/recipes/immutability-helpers.md

# Immutability Helpers

Redux is a tiny pattern that represents states as immutable objects. Redux was originally designed for React. Most Redux concepts, such as pure functions, are centered around the React ecosystem. Nowadays Redux is not directly related to React.

The cornerstone of Redux is immutability. Immutabilty is an amazing pattern to minimise unpredictable behaviour in our code. We're not going to cover functional programming in this article. However we're going to look at very useful packages that are called "immutability helpers". Most developers have to deal with, so called, "deep objects" and most important follow the immutability concept, when it comes to changing the value of some deeply nested property.

> 

Reviewer: We should add our own state operators here at the top of this list with a link to the page on that.
State Operators are essentially immutibility helpers.

Also, we could add https://github.com/kolodny/immutability-helper to the list below. I have used it before and it works well. It is also quite well used for Redux.

Author: Good point

## immer

`immer` is the most popular library that allows you to work with immutable objects in a more convenient way. Given the following code:

```ts
export interface Task {
  title: string;
  dates: {
    startDate: string;
    dueDate: string;
  };
}

export interface TrelloStateModel {
  tasks: {
    [taskId: string]: Task;
  };
}

@State<TrelloStateModel>({
  name: 'trello',
  defaults: {
    tasks: {}
  }
})
export class TrelloState {}
```

Let's imagine that we're faced with the task of changing the `dueDate` property:

```ts
export class UpdateDueDate {
  static readonly type = '[Trello] Update due date';
  constructor(public taskId: string, public dueDate: string) {}
}
```

Let's see how we would implement the `updateDueDate` action handler:

```ts
export class TrelloState {

  @Action(UpdateDueDate)
  updateDueDate(ctx: StateContext<TrelloStateModel>, action: UpdateDueDate) {
    ctx.setState(state => ({
      tasks: {
        ...state.tasks,
        [action.taskId]: {
          ...state.tasks[action.taskId],
          dates: {
            ...state.tasks[action.taskId].dates,
            dueDate: action.dueDate
          }
        }
      }
    }));
  }

}

> ```

Reviewer: This section above that sets out the example scenario and showcases the problem of doing a deep update should really be in its own section titled "The Problem".
Then below this would be a section called "Some Solutions" under which each of the tools are showcased.

Such code is complicated to maintain and accompany. It's not self-descriptive and will be a problem for upcoming developers. Let's see how we could re-write this code with the help of `immer`:

```ts
import { produce } from 'immer';

export class TrelloState {

  @Action(UpdateDueDate)
  updateDueDate(ctx: StateContext<TrelloStateModel>, action: UpdateDueDate) {
    const state = produce(ctx.getState(), draft => {
      draft.tasks[action.taskId].dates.dueDate = action.dueDate;
    });

    ctx.setState(state);
  }

}
```

Oh, have you noticed it's less code and looks much better. From the `immer` repisitory:
> Using Immer is like having a personal assistant; he takes a letter (the current state) and gives you a copy (draft) to jot changes onto. Once you are done, the assistant will take your draft and produce the real immutable, final letter for you (the next state).

> 

Reviewer: In addition to the example above I think that we should show that you can also use immer's `produce` function as a state operator like this:
```
import { produce } from 'immer';
export class TrelloState {
  @Action(UpdateDueDate)
  updateDueDate(ctx: StateContext<TrelloStateModel>, action: UpdateDueDate) {
    ctx.setState(produce(draft => {
      draft.tasks[action.taskId].dates.dueDate = action.dueDate;
    }));
  }
}
```

Author: OK, I will add 2 ways of using `immer`

```ts
export interface Task {
  title: string;
  dates: {
    startDate: string;
    dueDate: string;
  };
}

export interface TrelloStateModel {
  tasks: {
    [taskId: string]: Task;
  };
}

@State<TrelloStateModel>({
  name: 'trello',
  defaults: {
    tasks: {}
  }
})
export class TrelloState {}
```

Let's imagine that we're faced with the task of changing the `dueDate` property:

```ts
export class UpdateDueDate {
  static readonly type = '[Trello] Update due date';
  constructor(public taskId: string, public dueDate: string) {}
}
```

Let's see how we would implement the `updateDueDate` action handler:

```ts
export class TrelloState {

  @Action(UpdateDueDate)
  updateDueDate(ctx: StateContext<TrelloStateModel>, action: UpdateDueDate) {
    ctx.setState(state => ({
      tasks: {
        ...state.tasks,
        [action.taskId]: {
          ...state.tasks[action.taskId],
          dates: {
            ...state.tasks[action.taskId].dates,
            dueDate: action.dueDate
          }
        }
      }
    }));
  }

}
```

> Although this code will work but unfortunately it's complicated to maintain and accompany. It's not self-descriptive and will be daunting for upcoming developers.

Reviewer: ```suggestion
This code will work but unfortunately it is complicated to maintain and understand. It's not self-descriptive and will be daunting for new developers.
```

## Solutions

> There are different ways to stay away from such code. This can be accomplished with the use of different packages that may be suitable for your needs.

Reviewer: ```suggestion
There are different ways to improve this code. Let us look at a few different packages that can help in this regard.
```

### State Operators

> [State operators](../advanced/operators.md) are the first immutability helpers that NGXS provides out of the box. The `patch` operator will become your best friend in case of choosing state operators as your immutability helpers. Let's see how we could re-write the above code by dint of the `patch` operator:

Reviewer: ```suggestion
[State operators](../advanced/operators.md) are first-class immutability helpers that NGXS provides out of the box. The `patch` operator will become your best friend in case of choosing state operators as your immutability helpers. Let's see how we could re-write the above code with the help of the `patch` state operator:
```

```ts
import { patch } from '@ngxs/store/operators';

export class TrelloState {

  @Action(UpdateDueDate)
  updateDueDate(ctx: StateContext<TrelloStateModel>, action: UpdateDueDate) {
    ctx.setState(
      patch({
        tasks: patch({
          [action.taskId]: patch({
            dates: patch({
              dueDate: action.dueDate
            })
          })
        })
      })
    );
  }

}
```

### immer

> `immer` is the most popular library that allows you to work with immutable objects in a more convenient way. The below code shows how to write the same code with the help of Immer:

Reviewer: ```suggestion
`immer` is a very popular library that allows you to make changes to immutable objects as if they were mutable. The below code shows how to write the same code with the help of Immer:
```

```ts
import { produce } from 'immer';

export class TrelloState {

  @Action(UpdateDueDate)
  updateDueDate(ctx: StateContext<TrelloStateModel>, action: UpdateDueDate) {
    const state = produce(ctx.getState(), draft => {
      draft.tasks[action.taskId].dates.dueDate = action.dueDate;
    });

    ctx.setState(state);
  }

}
```

Immer's `produce` function can be also used as a state operator:

```ts
import { produce } from 'immer';

export class TrelloState {

  @Action(UpdateDueDate)
  updateDueDate(ctx: StateContext<TrelloStateModel>, action: UpdateDueDate) {
    ctx.setState(
      produce(draft => {
        draft.tasks[action.taskId].dates.dueDate = action.dueDate;
      })
    );
  }

}
```

> Oh, have you noticed it's less code and looks much better. From the `immer` repisitory:

Reviewer: ```suggestion
You may notice how much less code this is and how much better it looks. From the `immer` repository:
```

> Using Immer is like having a personal assistant; he takes a letter (the current state) and gives you a copy (draft) to jot changes onto. Once you are done, the assistant will take your draft and produce the real immutable, final letter for you (the next state).

[Immer repository](https://github.com/immerjs/immer)

### immutability-helper

> `immutability-helper` is a petty package that lets you mutate a copy of data without changing the original source:

Reviewer: ```suggestion
`immutability-helper` is a small package that lets you mutate a copy of data without changing the original source:
```

```ts
import update from 'immutability-helper';

export class TrelloState {

  @Action(UpdateDueDate)
  updateDueDate(ctx: StateContext<TrelloStateModel>, action: UpdateDueDate) {
    const state = update(ctx.getState(), {
      tasks: {
        [action.taskId]: {
          dates: {
            dueDate: {
              $set: action.dueDate
            }
          }
        }
      }
    });

    ctx.setState(state);
  }

}
```

> [immutability-helper repisitory](https://github.com/kolodny/immutability-helper)

Reviewer: ```suggestion
[immutability-helper repository](https://github.com/kolodny/immutability-helper)
```

### object-path-immutable

`object-path-immutable` is a small library that allows you to modify deep object properties without modifying the original object. Let's look at how we could write the same code using this library:

```ts
import immutable from 'object-path-immutable';

export class TrelloState {

  @Action(UpdateDueDate)
  updateDueDate(ctx: StateContext<TrelloStateModel>, action: UpdateDueDate) {
    const state = immutable.set(
      ctx.getState(),
      `tasks.${action.taskId}.dates.dueDate`,
      action.dueDate
    );

    ctx.setState(state);
  }

}
```

[object-path-immutable repository](https://github.com/mariocasciaro/object-path-immutable)

### immutable-assign

> `immutable-assign` is a lightweight library that pursues the same goal. Its syntax is alike to `immer`'s:

Reviewer: ```suggestion
`immutable-assign` is a lightweight library that pursues the same goal. Its syntax is similar to `immer`'s:
```

```ts
import * as iassign from 'immutable-assign';

export class TrelloState {

  @Action(UpdateDueDate)
  updateDueDate(ctx: StateContext<TrelloStateModel>, action: UpdateDueDate) {
    const state = iassign(ctx.getState(), state => {
      state.tasks[action.taskId].dates.dueDate = action.dueDate;
      return state;
    });

    ctx.setState(state);
  }

}
```

[immutable-assign repository](https://github.com/engineforce/ImmutableAssign)

### Ramda

> Ramda is an adorable library for functional programming and it's used in a large number of projects. This example might be useful for people who use both Ramda and NGXS in their projects:

Reviewer: ```suggestion
Ramda is a great library for functional programming and it is used in a large number of projects. This example might be useful for people who use both Ramda and NGXS in their projects:
```
